linkerd/stack/src/switch_ready.rs

```diff
@@ -0,0 +1,389 @@
+use super::NewService;
+use futures::future::Either;
+use std::{
+    future::Future,
+    pin::Pin,
+    task::{Context, Poll},
+    time::Duration,
+};
+use tokio::time::{delay_for, Delay, Instant};
+/// A service which falls back to a secondary service if the primary service
+/// takes too long to become ready.
+#[derive(Debug)]
+pub struct SwitchReady<A, B> {
+    primary: A,
+    secondary: B,
+    switch_after: Duration,
+    delay: Delay,
+    state: State,
+}
+
+#[derive(Debug, Clone)]
+pub struct NewSwitchReady<A, B> {
+    new_primary: A,
+    new_secondary: B,
+    switch_after: Duration,
+}
+
+#[derive(Copy, Clone, Eq, PartialEq, Debug)]
+enum State {
+    Primary,
+    Waiting,
+    Secondary,
+}
+
+// === impl NewSwitchReady ===
+
+impl<A, B> NewSwitchReady<A, B> {
+    /// Returns a new `NewSwitchReady`.
+    ///
+    /// This will forward requests to the primary service, unless it takes over
+    /// `switch_after` duration to become ready. If the duration is exceeded,
+    /// the `secondary` service is used until the primary service becomes ready again.
+    pub fn new(new_primary: A, new_secondary: B, switch_after: Duration) -> Self {
+        Self {
+            new_primary,
+            new_secondary,
+            switch_after,
+        }
+    }
+}
+
+impl<A, B, T> NewService<T> for NewSwitchReady<A, B>
+where
+    T: Clone,
+    A: NewService<T>,
+    B: NewService<T>,
+{
+    type Service = SwitchReady<A::Service, B::Service>;
+
+    fn new_service(&mut self, target: T) -> Self::Service {
+        SwitchReady::new(
+            self.new_primary.new_service(target.clone()),
+            self.new_secondary.new_service(target.clone()),
+            self.switch_after,
+        )
+    }
+}
+
+// === impl SwitchReady ===
+
+impl<A, B> SwitchReady<A, B> {
+    /// Returns a new `SwitchReady`.
+    ///
+    /// This will forward requests to the primary service, unless it takes over
+    /// `switch_after` duration to become ready. If the duration is exceeded,
+    /// the `secondary` service is used until the primary service becomes ready again.
+    pub fn new(primary: A, secondary: B, switch_after: Duration) -> Self {
+        Self {
+            primary,
+            secondary,
+            switch_after,
+            // the delay is reset whenever the service becomes unready; this
+            // initial one will never actually be used, so it's okay to start it
+            // now.
+            delay: delay_for(switch_after),
+            state: State::Primary,
+        }
+    }
+}
+
+impl<A, B, R> tower::Service<R> for SwitchReady<A, B>
+where
+    A: tower::Service<R>,
+    B: tower::Service<R, Response = A::Response, Error = A::Error>,
```

Comment:
note that if you use `tower::util::Either`, you don't need to require that the error types are the same (though they both need to be `Into<Error>`)

Reply:
didn't realize `Either` was `Into<Error>` when both arms are `Into<Error>`...mainly i thought i'd avoid adding a `map_err` here that could be added in a separate layer?

Reply:
oh, i see, it has a future impl, that's neat